internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+	defer lockmanager.ReadUnlock(lockKey)
+
+	return this.RemainingItems
+}
+
+func (this JobOutput[InputType, OutputType]) GetRunTime() int64 {
+	lockmanager.ReadLock(this.JobID)
+	defer lockmanager.ReadUnlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_RUN_TIME)
+	lockmanager.ReadLock(lockKey)
+	defer lockmanager.ReadUnlock(lockKey)
+
+	return this.RunTime
+}
+
+func (this JobOutput[InputType, OutputType]) SetError(err error) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_ERROR)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.Error = err
+}
+
+func (this JobOutput[InputType, OutputType]) SetWorkErrors(workErrors map[int]error) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_WORK_ERRORS)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.WorkErrors = workErrors
+}
+
+func (this JobOutput[InputType, OutputType]) SetResultItems(resultItems []OutputType) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_RESULT_ITEMS)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.ResultItems = resultItems
+}
+
+func (this JobOutput[InputType, OutputType]) SetRemainingItems(remainingItems []InputType) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_REMAINING_ITEMS)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.RemainingItems = remainingItems
+}
+
+func (this JobOutput[InputType, OutputType]) SetRunTime(runTime int64) {
+	lockmanager.Lock(this.JobID)
+	defer lockmanager.Unlock(this.JobID)
+
+	lockKey := fmt.Sprintf(JOB_OUTPUT_LOCK_FORMAT, this.JobID, JOB_OUTPUT_RUN_TIME)
+	lockmanager.Lock(lockKey)
+	defer lockmanager.Unlock(lockKey)
+
+	this.RunTime = runTime
+}
+
+// Given a customized Job, Job.Run() processes input items in a parallel pool of workers.
+// Returns the collected results in a JobOutput.
+// If the context is canceled during execution, returns nil.
+// When not waiting for completion, the JobOutput may still be modified until the Done channel is closed.
+// Returning a copy of results gives immediate access to stable results but the final results will not be accessible later.
+func (this *Job[InputType, OutputType]) Run() *JobOutput[InputType, OutputType] {
+	done := make(chan any)
+
+	output := JobOutput[InputType, OutputType]{
+		Done:           done,
+		ResultItems:    make([]OutputType, 0, len(this.WorkItems)),
+		RemainingItems: this.WorkItems,
+		RunTime:        0,
+		WorkErrors:     make(map[int]error, 0),
+		JobID:          util.UUID(),
+	}
+
+	err := this.Validate()
+	if err != nil {
+		close(done)
```

Comment:
I would put your closes right up against the return,
to make it clear that you are doing this on every return except the final one.